compiler/noirc_frontend/src/elaborator/types.rs

```diff
@@ -183,24 +182,16 @@ impl<'context> Elaborator<'context> {
             _ => (),
         }
 
-        // Check that any types with a type kind match the expected type kind supplied to this function
-        // TODO(https://github.com/noir-lang/noir/issues/5156): make this named generic check more general with `*resolved_kind != kind`
-        // as implicit numeric generics still existing makes this check more challenging to enforce
-        // An example of a more general check that we should switch to:
-        // if resolved_type.kind() != kind.clone() {
-        //     let expected_typ_err = CompilationError::TypeError(TypeCheckError::TypeKindMismatch {
-        //         expected_kind: kind.to_string(),
-        //         expr_kind: resolved_type.kind().to_string(),
-        //         expr_span: span.expect("Type should have span"),
-        //     });
-        //     self.errors.push((expected_typ_err, self.file));
-        //     return Type::Error;
-        // }
-        if let Type::NamedGeneric(_, name, resolved_kind) = &resolved_type {
-            if matches!(resolved_kind, Kind::Numeric { .. }) && matches!(kind, Kind::Normal) {
+        if let Type::NamedGeneric(_type_var, _name, resolved_kind) = &resolved_type {
+            assert_eq!(&resolved_type.kind(), resolved_kind);
+            if resolved_type.kind() != kind.clone() {
```

Comment:
```suggestion
            if resolved_type.kind() != *kind {
```
Don't need to clone since `!=` should reference its arguments